sqlmesh/core/engine_adapter/base.py

```diff
@@ -227,12 +241,31 @@ def _get_source_queries(
         target_table: TableName,
         *,
         batch_size: t.Optional[int] = None,
+        source_columns: t.Optional[t.List[str]] = None,
     ) -> t.List[SourceQuery]:
         import pandas as pd
 
         batch_size = self.DEFAULT_BATCH_SIZE if batch_size is None else batch_size
-        if isinstance(query_or_df, (exp.Query, exp.DerivedTable)):
-            return [SourceQuery(query_factory=lambda: query_or_df)]  # type: ignore
+        if isinstance(query_or_df, exp.Query):
+            query_factory = lambda: query_or_df
+            if source_columns:
+                if not columns_to_types:
+                    raise SQLMeshError("columns_to_types must be set if source_columns is set")
+                if not set(columns_to_types).issubset(set(source_columns)):
```

Comment:
Perhaps this is just a syntax issue but I have a hard time interpreting why would `columns_to_types` ever be a subset of `source_columns`.

Reply:
Well this logic doesn't handle the case currently when source columns is more (which will come with ignore additive  changes). Its strictly written to deal with when `columns_to_types` is not a subset.